core/kazoo_auth/src/kz_mfa_auth.erl

```diff
@@ -0,0 +1,111 @@
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2017, 2600Hz, INC
+%%% @doc
+%%%
+%%% @end
+%%% @contributors
+%%%-------------------------------------------------------------------
+-module(kz_mfa_auth).
+
+-export([authenticate/1
+        ,get_configs/1
+        ,get_system_configs/0
+        ,provider/1
+        ]).
+
+-include("kazoo_auth.hrl").
+
+-type result() :: mfa_result().
+
+-export_type([result/0]).
+
+%%--------------------------------------------------------------------
+%% @public
+%% @doc Read configuration and do authentication with configured MFA provider
+%% @end
+%%--------------------------------------------------------------------
+-spec authenticate(kz_proplist()) -> result().
+authenticate(Claims) ->
+    Configs = get_configs(props:get_value(<<"mfa_options">>, Claims)),
+    case provider(Configs) of
+        {'disabled', _Provider} ->
+            ErrorMsg = io_lib:format("provider ~s is disabled", [_Provider]),
+            lager:debug(ErrorMsg),
+            {'error', kz_term:to_binary(ErrorMsg)};
+        ?NE_BINARY=Provider ->
+            Module = module_name(Provider),
+            lager:debug("performing authentication factor with ~s(~p) provider"
+                       ,[Provider, Module]
+                       ),
+            Module:authenticate(Claims, kz_json:get_value(<<"settings">>, Configs, kz_json:new()));
+        _ ->
+            lager:debug("no provider is available or configured"),
+            {'error', 'no_provider'}
+    end.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc Get MFA provider and checks it's enabled or not
+%% @end
+%%--------------------------------------------------------------------
+-spec provider(api_object()) -> ne_binary() | {'disabled', ne_binary()} | {'error', 'no_provider'}.
+provider(Configs) ->
+    Name = kz_json:get_value(<<"provider_name">>, Configs),
+    case kz_json:is_true(<<"enabled">>, Configs, 'true') of
+        'true' when ?NE_BINARY =/= Name -> Name;
```

Comment:
how about using is_empty instead of a matching macro which I'm surprised even compiles?
how does this even work?? wow

yeah no that should be a compile error

```erlang
1> <<_:8,_/binary>> = <<"bla">>.
<<"bla">>
2> <<_:8,_/binary>> == <<"bla">>.
* 1: variable '_' is unbound
3> <<_:8,_/binary>> =/= <<"bla">>.
* 1: variable '_' is unbound
```